tests/test_metadata/test_v2.py

```diff
@@ -316,3 +316,28 @@ def test_zstd_checksum() -> None:
         arr.metadata.to_buffer_dict(default_buffer_prototype())[".zarray"].to_bytes()
     )
     assert "checksum" not in metadata["compressor"]
+
+
+@pytest.mark.parametrize(
+    "fill_value", [None, np.void((0, 0), np.dtype([("foo", "i4"), ("bar", "i4")]))]
+)
+def test_structured_dtype_fill_value_serialization(tmp_path, fill_value):
+    group_path = tmp_path / "test.zarr"
+    root_group = zarr.open_group(group_path, mode="w", zarr_format=2)
+    dtype = np.dtype([("foo", "i4"), ("bar", "i4")])
+    root_group.create_array(
+        name="structured_dtype",
+        shape=(100, 100),
+        chunks=(100, 100),
+        dtype=dtype,
+        fill_value=fill_value,
+    )
+
+    zarr.consolidate_metadata(root_group.store, zarr_format=2)
+    root_group = zarr.open_group(group_path, mode="r")
+    assert (
+        root_group.metadata.consolidated_metadata.to_dict()["metadata"]["structured_dtype"][
+            "fill_value"
+        ]
+        == fill_value
```

Comment:
@ilan-gold I'm a little confused by this test. If the fill value is a structured dtype scalar, then shouldn't the fill value that appears in metadata be base64 encoded? If so, shouldn't this check fail in that case?

Reply:
I think the python representation here is serlialized into the void data type.  Wheter that is correct or not is a different story.  Looking at it closely, I think this is either (a) expected, and the typing on `ArrayV2Metadata` is wrong or (b) the typing is right, and the behavior is wrong.

The type is: `fill_value: int | float | str | bytes | None = 0`

But the call to `parse_fill_value` yields `numpy` object:

https://github.com/zarr-developers/zarr-python/blob/36a1bac850c50fbaa1708f8986edd5766f5fd67b/src/zarr/core/metadata/v2.py#L297-L308

Reply:
(What I'm trying to say is that this dictionary is not the on-disk json, but a parsed version)

Reply:
Definitely not offhand, but just to understand, why does it matter? Is there a backwards compat concern with the in-memory python representation? TBH structured data types are much less essential to us than some of the other people who are raising these concerns from what it sounds like so I'm not super familiar with previous behavior.  I don't think many people in our community use them, but they are in our CI and I like contributing so I make these PRs :)

Reply:
I'm touching a lot of data type representation code over in #2874 and I want to make sense of some of the test failures I'm seeing, and this was one of the tests that I tripped. I think your explanation makes sense (i.e., this is just the in-memory representation, and so the fill value should be the decoded version), sorry for the noise!

Reply:
No worries @d-v-b always happy to help.  Will be quick to report on the status of this all once 3.0.8 is out, but our tests show no errors with this feature at the moment.

Reply:
for posterity, the specific thing in my PR that made this test fail was the use of `to_dict`. In #2874, I am making fill value encoding happen in the call to `to_dict`, instead via a special JSON encoder. So on my branch this test was comparing the JSON-serialized fill value against the in-memory version. I made the test pass by removing the `to_dict` step and directly comparing the `metadata.fill_value` attribute against the expected `fill_value`.